Streaming filters

ethomson opened the pull request:
This is an attempt at providing a streaming functionality for filters.  Generally speaking, the streaming interface is fairly straightforward:  we set up a chain of filters such that each writes into the next filter, and the final filter will use the same interface to write into the actual writing target (something that puts it into the ODB or the working directory).  A consumer will ask the `filter_list` for a stream, and then will simply write to it as it sees fit, and that filter will be responsible for doing a transformation and then passing the data along to the next stream.

I have updated the internals of the library (checkout, for example) to use the new streams.  Filters using the existing `apply` API are still supported (we create a "proxy" stream that loads the entire data and then passes it to apply).  

There is also a test that optionally (when the `GITTEST_INVASIVE_FILESYSTEM` environment variable is set) pushes a 500 MB file through filters that compress it into the ODB (and decompress it into the working directory) so that you can see that filters are operational without loading the entire file into RAM.

Two things I want to call out, though I'm certain that reviews will find more:
1. Right now the ultimate writer uses the same interface as the streaming filter.  So checkout, for instance, passes a `git_filter_stream` itself, which is the terminal writer, not an actual pass-through stream.  This is semantically a little weird, I think.  It might make more sense to simply have another base with `write` and `close` semantics (and presumably `git_filter_stream` would extend *this*) though we would have to trade the additional layer of abstraction versus the naming.  Feedback is welcome.
2. I did not change the existing filters (ident, CR/LF) to stream.  Streaming CR/LF filters would actually break the existing semantics, as they load the entire file to do the stats.  This means that if you had poorly configured your .gitattributes then you might try to CR/LF filter some giant file and we would OOM.  We could either:  change the CR/LF filter to only perform stats on the first `n` bytes (which seems very reasonable) or have some ability to abort and rewind the filters (which seems difficult, to say the least.)

carlosmn commented:
git's crlf filter is streaming, so it's possible to keep git's semantics that way, though I don't know off-hand how much of a stream it is. Aborting might be doable if we can recognise that there's a single filter in effect, so we can replay the stream from a blob or file, but we don't have to worry about that right now.

carlosmn commented:
As for the checkout stream being a `filter_stream`, this might just come down to naming. The interface you're defining there is just a `WriteStream` or `WriteCloser` (depending on your favourite language this week) so I don't see an issue with checkout defining a sink which writes out to disk with this interface. Maybe if you named it `git_writestream` it'd look just fine.

carlosmn commented on src/checkout.c:
```diff
@@ -1371,39 +1372,37 @@ static int mkpath2file(
-	if ((error = mkpath2file(data, path, data->opts.dir_mode)) < 0)
-		return error;
+struct checkout_stream {
+	git_filter_stream base;
```
Just a nitpick, but we do have a preference for calling this 'parent', rather than base (I think it's just the iterators which use this name).

ethomson replied:
Fixed.

carlosmn commented on src/checkout.c:
```diff
@@ -1411,36 +1410,83 @@ static int blob_content_to_file(
+	assert(writer.open == 0);
 
 	git_filter_list_free(fl);
+	p_close(fd);
```
Why perform the close here instead of giving ownership of the descriptor to the filter and have its `->close()` close the descriptor.

Here you're also throwing away the return value from close, which can return an error (though I'm sure we do this in very many other places).

ethomson replied:
Yes, that's much better.  Fixed; thanks.

carlosmn commented on src/checkout.c:
```diff
@@ -1411,36 +1410,83 @@ static int blob_content_to_file(
+		return error;
+
+	if (fl)
+		git_filter_list__set_temp_buf(fl, &data->tmp);
```
This `if` looks odd. Can we *not* have the filter list here? We seem to assume later on that we do have it.

ethomson replied:
We can not - the existing filters system let you pass a `NULL` `filter_list` to simply skip the filtering.  (https://github.com/libgit2/libgit2/blob/master/src/filter.c#L631)  This maintains that.

carlosmn replied:
Fair enough, I hadn't noticed that we always use functions around `fl` and do check if it's `NULL` each time.

ethomson commented:
Yep, I changed this to `git_writestream`.  I was thinking that there was something in `git_filter_stream` that was unique to filters.  There's not.  I dropped it in favor of `git_writestream`.  Thanks!

carlosmn commented:
The existence of `git_filter_list__set_temp_buf()` seems a bit suspicious. Why is there a per-filte_list buffer which we set from just the one place? Can we instead pass it as an argument somewhere?

I think we should extract `filter_list_out_buffer_from_raw()` as `git_buf_attach_nonowned()` or a better name, as it's not about filters but making a `git_buf` point to memory which outlives it. We don't have to do it in this PR but there was another PR where the code does this manually and it would help readability if we could delegate to a function.

We should activate `GITTEST_INVASIVE_FILESYSTEM` on travis.

Other than the buffer thing, the code looks pretty good. We should merge this in soon so we can see how the bindings can work with it.

ethomson commented:
Yeah, I zapped `set_temp_buf` and instead fleshed out a more libgit2-idiomatic `filter_list__load_ext` that takes a `filter_options` struct.  (In doing so, I renamed `git_libgit2_opt_t` to `git_libgit2_flag_t` to better match our pattern of "options" being structs.)

The only really breaking change here is changing `GIT_FILTER_OPT_DEFAULT` -> `GIT_FILTER_DEFAULT`, which I can revert if that seems needlessly gratuitous.

carlosmn commented:
:heart_eyes:

I don't feel bad about breaking the API if we're making it more consistent. Unless someone has objections, I'm going to merge this soon.
